### DISCUSSION TITLE
Should the `Parser.set` function validate variable names?

### DISCUSSION THREAD

**Discussion by orelbn** [josdejong/mathjs#3290] [Design decisions]
According to documentation, variable names must meet certain criteria:  
https://mathjs.org/docs/expressions/syntax.html#constants-and-variables
![image](https://github.com/user-attachments/assets/d7d90f89-3cdb-4bc6-943c-e42c989a7125)

But if you use the set function on the parser, it does not validate that criteria.
  
  ```
  /**
   * Set a symbol (a function or variable) by name from the parsers scope.
   * @param {string} name
   * @param {* | undefined} value
   */
  Parser.prototype.set = function (name, value) {
    this.scope.set(name, value)
    return value
  }
 ```

**Comment by josdejong** [comment]
Thanks, good point. Right now the parser allows setting any variable name, and only when using the variable in an expression, an error will be thrown. It would be good to validate the variable name in the method `Parser.set`. 

@nkumawat34 this actually is a very nice first issue, is this something you would like to pick up?

**Comment by nkumawat34** [comment]
Yes I will pick it.

**Reply by josdejong** [reply]
Thanks! 💪

**Comment by nkumawat34** [comment]
```
  function isValidVariableName(name) {
    const reservedKeywords = ['mod', 'to', 'in', 'and', 'xor', 'or', 'not', 'end'];

    // Regular expression to match the variable name rules
    const regex = /^[a-zA-Z_$\u00C0-\u02AF\u0370-\u03FF\u2100-\u214F\u{1D400}-\u{1D7FF}][a-zA-Z0-9_$\u00C0-\u02AF\u0370-\u03FF\u2100-\u214F\u{1D400}-\u{1D7FF}]*$/u;

    // Check if the name matches the regex and is not a reserved keyword
    return regex.test(name) && !reservedKeywords.includes(name);
  }

  /**
   * Set a symbol (a function or variable) by name from the parsers scope.
   * @param {string} name
   * @param {* | undefined} value
   */
    Parser.prototype.set = function (name, value) {
      if (!isValidVariableName(name)) {
        throw new Error(`Invalid variable name: '${name}'. Variable names must follow the specified rules.`);
      }
      this.scope.set(name, value);
      return value;
    };

**Reply by nkumawat34** [reply]
This code I have written for variable checking .

**Reply by josdejong** [reply]
I think it is important that the rules exactly match what the expression parser allows. So, the characters of the variable name should be checked with the [`parse.isAlpha`](https://github.com/josdejong/mathjs/blob/develop/src/expression/parse.js#L501-L505) and `parse.isDigit` functions I think:

I think we can implement `isValidVariableName` by copying the logic used in the the expression parser to parse a variable name (loop over all characters and check whether the character is valid):

https://github.com/josdejong/mathjs/blob/982bbdcb9b4513e4c00db7e8f09e361eeebec678/src/expression/parse.js#L433-L437

Also, the expression parser allows to use operators like `mod` as variable names, so `Parser` should allow that too, does that make sense?

**Reply by nkumawat34** [reply]
Could you assist me a little? I’ve attempted to use functions like parse.isAlpha and parse.isDigit in my Parser.js file for creating rules by importing them, but I can’t access them. Do I need to define those functions in the Parser.js file to use them?

**Reply by josdejong** [reply]
You should be able to inject `parse` into `Parser` and then use those functions, something like this:

```js
const name = 'Parser'
const dependencies = ['evaluate', 'parse']

export const createParserClass = /* #__PURE__ */ factory(name, dependencies, ({ evaluate, parse }) => {
  // ...

  function isValidVariableName(name) {
    // TODO: loop over all characters of name and test with parse.isAlpha and parse.isDigit
  }
  
  return Parser
}, { isClass: true })
```

**Reply by nkumawat34** [reply]
```
 function isValidVariableName(name) {
    // Iterate over each character in the name
    for (let i = 0; i < name.length; i++) {
      const c = name[i]; // Current character
      const cPrev = i > 0 ? name[i - 1] : null; // Previous character or null if there isn't one
      const cNext = i < name.length - 1 ? name[i + 1] : null; // Next character or null if there isn't one
  
      if (i === 0) {
        // First character must be alpha character
        if (!parse.isAlpha(c, cPrev, cNext)) {
          return false;
        }
      } else {
        // Subsequent characters must be alphanumeric
        if (!(parse.isAlpha(c, cPrev, cNext) || parse.isDigit(c))) {
          return false;
        }
      }
    }
    return true; // Name is valid
  }

**Reply by nkumawat34** [reply]
i think now it looks good

**Reply by josdejong** [reply]
Yes indeed 👍 I think the function can be simplified a bit but th logic is exactly what we need.

**Reply by nkumawat34** [reply]
I thought to simplify more this function but I am not getting any idea to do it.

**Reply by josdejong** [reply]
I think the following is a bit simpler. Using `charAt` instead of string index ensures that the characters are an empty string instead of `undefined` when out of range.

```js
  function isValidVariableName(name) {
    for (let i = 0; i < name.length; i++) {
      const cPrev = name.charAt(i - 1)
      const c = name.charAt(i)
      const cNext = name.charAt(i + 1)
      const valid = parse.isAlpha(c, cPrev, cNext) || (i > 0 && parse.isDigit(c))

      if (!valid) {
        return false
      }
    }

    return true
  }
```

**Reply by nkumawat34** [reply]
can I make a PR of this with unit test cases.

**Reply by josdejong** [reply]
Yes please 👍

**Reply by nkumawat34** [reply]
Please check the PR for issue https://github.com/josdejong/mathjs/pull/3308

**Reply by josdejong** [reply]
Thanks, the PR looks good on first sight. I'll look into it in more detail this week.

**Reply by josdejong** [reply]
A fix is published now in `mathjs@13.2.1`.
